### DISCUSSION TITLE
Possibility of offload token signing to an external service (e.g. AWS' KMS)

### DISCUSSION THREAD

**Discussion by paulolieuthier** [zitadel/zitadel#7434] [Ideas]
How hard would it be to make Zitadel use KMS to sign tokens? I would like not to manage any keys and offload all crypto operations.

**Comment by hifabienne** [comment]
@eliobischof @stebenz can you have a look at this?

**Reply by fforootd** [reply]
@paulolieuthier I wonder do you want to take the actually signing to JWT/SAML Tokens to the KMS or only the Key material lifecycle?

If zitadel needs to rely on an external source in the signing operation it needs for sure some configurations.
Like zitadel would need to know the type of keys and its signing algorithm to publish that info on the OIDC discovery/keys endpoint as well as the SAML metadata.

**Reply by paulolieuthier** [reply]
I would like to take it all to KMS. I guess in this scenario one could point Zitadel to a specific KMS key (managed by AWS) and it could retrieve (by using KMS API) all the necessary metadata.

**Reply by fforootd** [reply]
Got, it. Thanks for clarifying!

I think there would be two main concerns to implement:

1) Accessing KMS instead of the primary DB for key operations
2) Sending JWT and SAML token content to KMS for the signing operation

I think 2) would be the main effort and raises some questions. While quickly googling it looks like it will add 250-500ms in latency to token calls.

https://aws.amazon.com/kms/faqs/

<img width="486" alt="image" src="https://github.com/zitadel/zitadel/assets/9879976/797f6aa9-599b-49e8-bb0b-1413a39ab2ab">

**Reply by paulolieuthier** [reply]
Yes, we are aware of the added latency. It's unfortunate, but it's a trade-off we are comfortable with for security.

That quote, though, is about XKS, in which the keys are stored and used on-premises.

**Reply by fforootd** [reply]
> That quote, though, is about XKS, in which the keys are stored and used on-premises.

Haha, that could well be 😁 I am not exactly an AWS expert.
